### DISCUSSION TITLE
Security concern

### DISCUSSION THREAD

**Discussion by n3b0r** [livewire/livewire#2311] [1. Help] [answered]
Hi everyone,

I have a security concern about the render method. 

Let me explain with an example. My render method looks like:

```
public function render()
    {
        $this->authorize('isAdmin', auth()->user());
        
        return view('livewire.users', [
            'users' => User::with('role', 'clients')
                ->search(['name', 'email', 'clients.name', 'role.name'], $this->search)
                ->orderBy('name')
                ->paginate($this->paginate)
        ]);
    }
```

My question here is, it's needed to authorize the render method if you check the policy on web.php routes? Btw, this example is the typical section only allowed for the admin role. 

Thanks!

**Comment by nuernbergerA** [comment]
If you using it as [page component](https://laravel-livewire.com/docs/2.x/rendering-components#page-components) you can simply [protect your route](https://laravel.com/docs/8.x/authentication#specifying-a-guard).

if you using it `<livewire:some-component />` just wrap it with `@if` or `@can`

**Reply by n3b0r** [reply]
It is a Full-Page component, I can't do that ;)

**Reply by nuernbergerA** [reply]
What exactly is it that you cant do?

**Reply by n3b0r** [reply]
That is a more complete Livewire component version. I'm trying to implement a user impersonation functionality. If I use auth on render, as I had before to implement the impersonation, all it's working perfectly. 

The problem appears after the user clicks on impersonation, due to before the redirect happens, the render method it's called and the impersonated user can't access to this section, so 403 error appears and the redirect doesn't happen.

```
<?php

namespace App\Http\Livewire;

use App\Models\User;
use Illuminate\Foundation\Auth\Access\AuthorizesRequests;
use Livewire\Component;
use Livewire\WithPagination;

class Users extends Component
{
    use WithPagination, AuthorizesRequests;
    
    //Model
    public User $user;
    
    //Table values & filters
    public $search = '';
    private $paginate = 10;
    
    //another stuff 
    
    public function impersonate(User $user)
    {
        session()->put('impersonate', auth()->user()->id);
        auth()->loginUsingId($user->id);
        
        return redirect('/');
    }
    
    public function render()
    {
        $this->authorize('isAdmin', auth()->user());
        
        return view('livewire.users', [
            'users' => User::with('role', 'clients')
                ->search(['name', 'email', 'clients.name', 'role.name'], $this->search)
                ->orderBy('name')
                ->paginate($this->paginate)
        ]);
    }
}

```

So, my concern here is, is it secure to avoid authorization on render method if I have /users route protected on web.php file?

I'm asking that because Livewire creates internal custom routes like http://path/livewire/messages/users to works.

**Reply by nuernbergerA** [reply]
All ` http://path/livewire/messages/*` endpoints are 100% secure, you can only interact if the fingerprint is equal (which changes on each request)

**Reply by n3b0r** [reply]
Thanks for your replays @nuernbergerA :)

The last question only to make it crystal clear.

If I do something like:

**web.pbp**
```
Route::middleware(['can:isAdmin,App\Models\User'])->group(function () {
        Route::get('users', Users::class)->name('users');
});
```

There is no needed to protect any method inside Livewire's component all http://path/livewire/messages/* routes are secured and no one can access them directly right? 

So the authorize method should be used on components that mix methods that could be accessed by different user roles.

**Reply by n3b0r** [reply]
Those questions come to me due to over tests side you can reach methods directly.

Let me show you an example:

```
/** @test * */
    public function random_test()
    {    
        Livewire::actingAs($unauthorized_user_role)
            ->call('save')
            ->assertForbidden()    
    } //it fails 
```

**Reply by n3b0r** [reply]
Hey, @nuernbergerA sorry for the ping. Did you check my last messages? I'm awaiting your reply to deploy that feature into the production branch ;)

Thanks in advance

**Reply by n3b0r** [reply]
Hi @joshhanley, please could you help us with our last question about security concerns about how Livewire manages internal routes access? 

I will be more than grateful. Thanks in advance

**Comment by joshhanley** [comment, marked as answer]
@vgsystems yeah as @nuernbergerA has said the internal routes would require some one calculating the fingerprint and the data checksum to be able to access them, and as the checksum is generated using your app key, that would be next to impossible without that key. If they have that key, then you have bigger problems 😆

From a testing perspective though, if you are using the `call` method in your test, that actually calls the method on the component itself, like a unit test on the component. It doesn't go through the router, hence why you don't get any route level authorisation.

If you only need one level of authorisation for that route and component, then I would just add a normal `get` test where you test the route with authorisation at the top of your testing file to ensure that the route is restricted.

But if different actions have different authorisation levels, then I would be putting authorise in each of the methods, in which you should be able to test that when you use `call` in your Livewire test.

Hope this helps!

**Reply by n3b0r** [reply]
Thank you so much for your reply @joshhanley now it's crystal clear.

**Reply by joshhanley** [reply]
No worries!
